### PR TITLE
API Endpoints Doc

### PR DESCRIPTION
**Pull request by mzhang2021**
@ahsu1230 
@tonywu315

### CODE REVIEW

File: resources/backend/api_endpoints.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# API Endpoints
+API endpoints are essentially points of communication in which a user can interact with and utilize an API's functions. In the context of this project, they are ways for frontend users to interact with and use backend functions.
+![API Endpoint Diagram](../images/api_endpoint_diagram.png)
```
Comment (mzhang2021) [inline resources/backend/api_endpoints.md:3]: The image won't show properly until this gets merged cause the url is different.

File: resources/backend/api_endpoints.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+Imagine the backend is an island with multiple ship ports. If a frontend user wants to interact with the backend, they send a ship to one of the ports to either deliver, retrieve, or modify information. These ships are called **HTTP requests**.
+## What does an endpoint look like?
```
Comment (ahsu1230) [inline resources/backend/api_endpoints.md:5]: Looks good! I like the island analogy! To be more clear (explicit), an incoming ship is an HTTP request while an outgoing ship is the HTTP response. Each of the ship ports is an HTTP endpoint.

File: resources/backend/api_endpoints.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+For example, the above example of retrieving a program with a certain id would use a GET request, since we are simply reading information and not modifying anything.
+## What do API endpoints return?
+### Status Codes
+Status codes are essentially abbreviated errors that an API endpoint could return. Here are some examples of common status codes:
+* 200 (OK) - the API successfully carried out what was requested.
+* 400 (Bad Request) - the request made was malformed in some way (invalid syntax, invalid parameters, etc.).
+* 404 (Not Found) - the API could not find any endpoint that matches with what was requested.
+* 500 (Internal Server Error) - something went wrong within the backend.
+### JSON Bodies
+JSON bodies can be sent both ways. A request could send a JSON body to the endpoint. For example, creating a program via the API endpoint `/api/programs/v1/create` requires the user to also send a JSON body of a program with the request.
```
Comment (ahsu1230) [inline resources/backend/api_endpoints.md:27]: Give a link to what JSON is. I think we have a doc explaining that.

File: resources/backend/api_endpoints.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+HTTP requests have multiple types, known as methods. A GET request retrieves information from an API, a POST request sends information to an API, and a DELETE request removes information from an API. There are other methods, but this project doesn't use them (yet).
+
+For example, the above example of retrieving a program with a certain id would use a GET request, since we are simply reading information and not modifying anything.
+## What do API endpoints return?
+### Status Codes
+Status codes are essentially abbreviated errors that an API endpoint could return. Here are some examples of common status codes:
+* 200 (OK) - the API successfully carried out what was requested.
+* 400 (Bad Request) - the request made was malformed in some way (invalid syntax, invalid parameters, etc.).
+* 404 (Not Found) - the API could not find any endpoint that matches with what was requested.
+* 500 (Internal Server Error) - something went wrong within the backend.
```
Comment (ahsu1230) [inline resources/backend/api_endpoints.md:25]: You can create a quick link to a list of all status codes on the Internet. I think Wikipedia details this?

File: resources/backend/api_endpoints.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+API endpoints can receive URI parameters. For example, if you want to access a program with id "prog1", the syntax for this specific API is `/api/programs/v1/program/prog1`. If you want to access the program with id "prog2", send a HTTP request to `/api/programs/v1/program/prog2`.
+
+API endpoints can also receive query parameters. For example, if you want to access all published programs from this sample API, you would use `/api/programs/v1/all?published=true`. You've probably seen query parameters in website URLs before. For example, if you search something in Google, the URL will look something like `https://www.google.com/search?source=hp&ei=DFKbXvqAIPCIytMP...`. The stuff after the question mark is a query parameter.
+
+We use URI parameters when we want to identify a specific resource(s), and we use query parameters as to sort or filter resources.
+## What does an HTTP request look like?
+HTTP requests have multiple types, known as methods. A GET request retrieves information from an API, a POST request sends information to an API, and a DELETE request removes information from an API. There are other methods, but this project doesn't use them (yet).
+
```
Comment (ahsu1230) [inline resources/backend/api_endpoints.md:17]: I would do a bullet point for this: The main types of HTTP requests are:
 - GET - retrieves ....
 - POST - ....
 - DELETE - ....

With explanations after each type. Or something like that?

File: resources/backend/api_endpoints.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+Status codes are essentially abbreviated errors that an API endpoint could return. Here are some examples of common status codes:
+* 200 (OK) - the API successfully carried out what was requested.
+* 400 (Bad Request) - the request made was malformed in some way (invalid syntax, invalid parameters, etc.).
+* 404 (Not Found) - the API could not find any endpoint that matches with what was requested.
+* 500 (Internal Server Error) - something went wrong within the backend.
+### JSON Bodies
+JSON bodies can be sent both ways. A request could send a JSON body to the endpoint. For example, creating a program via the API endpoint `/api/programs/v1/create` requires the user to also send a JSON body of a program with the request.
+
+An endpoint can also return a JSON body. For example, the aforementioned query for a program with a specific id returns a JSON body representing the program with that id in the database.
+## References/Extra Resources
```
Comment (ahsu1230) [inline resources/backend/api_endpoints.md:30]: Finally, can you add spaces in between every `##` so this doc is more readable? THANKS!

File: resources/backend/api_endpoints.md (hunk 3 above)
Comment (mzhang2021) [reply resources/backend/api_endpoints.md:46]: I actually couldn't find a doc explaining JSONs in the resources folder. I put in a w3schools link for now.

Comment (ahsu1230) [reply resources/backend/api_endpoints.md:46]: Ah I forgot about this. No file just on JSON, but there's a section here: https://github.com/ahsu1230/mathnavigatorSite/blob/master/resources/02_protocols.md#what-is-json

Do you mind linking it? You can merge the change yourself. Nbd

**Pull request merged**